fvm/errors/errors.go

```diff
@@ -97,3 +99,147 @@ func HandleRuntimeError(err error) error {
 	// All other errors are non-fatal Cadence errors.
 	return NewCadenceRuntimeError(runErr)
 }
+
+// TODO(patrick): combine ErrorCodeChecker, Error and Failure interface.
+// We can also combine CodedError and CodedFailure once the interface is
+// unified.
+type ErrorCodeChecker interface {
+	// TODO(patrick): add Code() ErrorCode
+	// TODO(patrick): add IsFailure() bool
+
+	HasErrorCode(code ErrorCode) bool
+
+	error
+}
+
+func HasErrorCode(err error, code ErrorCode) bool {
+	var checker ErrorCodeChecker
+	return As(err, &checker) && checker.HasErrorCode(code)
+}
+
+type codedError struct {
+	code ErrorCode
+
+	// The first error in the list is the root cause.
+	errors []error
+}
+
+func wrapError(
+	code ErrorCode,
+	rootCause error,
+) codedError {
+	// TODO(patrick): check if rootCause is a failure, and if so, invert the
+	// error nesting (convert the new code into an additional error).
+	return codedError{
+		code:   code,
+		errors: []error{rootCause},
+	}
+}
+
+func newError(
+	code ErrorCode,
+	format string,
+	formatArguments ...interface{},
+) codedError {
+	return wrapError(code, fmt.Errorf(format, formatArguments...))
+}
+
+func (err codedError) Error() string {
+	if len(err.errors) == 1 {
+		return fmt.Sprintf("%v %v", err.code, err.errors[0])
+	}
+	return fmt.Sprintf(
+		"%v: %s",
+		err.code,
+		multierror.ListFormatFunc(err.errors))
+}
+
+func (err codedError) Unwrap() error {
+	return err.errors[0]
```

Comment:
If we use `Caused` to append another error, then that error object seems won't be accessible, because `Unwrap` only returns the first error "errors[0]", is it useful to allow callers to iterate through other errors?